libs/wingsdk/src/simulator/simulator.ts

```diff
@@ -996,3 +987,66 @@ export interface SimulatorServerResponse {
   /** The error that occurred during the method call. */
   readonly error?: any;
 }
+
+/**
+ * Given the "current" set of resources and a "next" set of resources, calculate the diff and
+ * determine which resources need to be added, updated or deleted.
+ *
+ * Note that dependencies are not considered here but they are implicitly handled by the
+ * `startResource` and `stopResource` methods. So, for example, when a resource is updated,
+ * all of it's dependents will be stopped and started again.
+ */
+function planUpdate(current: BaseResourceSchema[], next: BaseResourceSchema[]) {
+  const currentByPath = toMap(current);
+  const nextByPath = toMap(next);
+
+  const added: string[] = [];
+  const updated: string[] = [];
+  const deleted: string[] = [];
+
+  for (const [path, nextConfig] of Object.entries(nextByPath)) {
+    const currConfig = currentByPath[path];
+
+    // if the resource is not in "current", it means it was added
+    if (!currConfig) {
+      added.push(nextConfig.path);
+      continue;
+    }
+
+    // the resource is already in "current", if it's different from "next", it means it was updated
+    const state = (r: BaseResourceSchema) =>
+      JSON.stringify({
+        props: r.props,
+        type: r.type,
+      });
+
+    if (state(currConfig) !== state(nextConfig)) {
+      updated.push(nextConfig.path);
+    }
+
+    // remove it from "current" so we know what's left to be deleted
+    delete currentByPath[path];
+  }
+
+  // everything left in "current" is to be deleted
+  for (const config of Object.values(currentByPath)) {
+    deleted.push(config.path);
+  }
+
+  return { added, updated, deleted };
+
+  function toMap(list: BaseResourceSchema[]): {
+    [path: string]: BaseResourceSchema;
+  } {
+    const ret: { [path: string]: BaseResourceSchema } = {};
+    for (const resource of list) {
+      if (ret[resource.path]) {
+        throw new Error(
+          `unexpected - duplicate resources with the same path: ${resource.path}`
+        );
+      }
+      ret[resource.path] = resource;
+    }
+    return ret;
+  }
```

Comment:
Super nit but if we are going to model dependencies then we could update the format of simulator.json to store the resources in an object/map instead of as an array like it is today

Reply:
I rather not break the schema at this point :-)